spec.md

```diff
@@ -1798,6 +1798,45 @@ module MyModule :
    out[4] <= in
 ```
 
+## Bit-indices
+
+The bit-index expression statically refers, by index, to a particular bit, or
+slice of bits, of an expression with an integer type (`UInt`{.firrtl} or
+`SInt`{.firrtl}). The indices must be non-negative integers and cannot be equal
+to or exceed the width of the integer they index. The bit-index `x[hi:lo]`
+selects bits `hi` (most significant) through `lo` (least significant) of `x`.
+The bit-index `x[i]` selects the single bit `i`.
+
+The type of the bit-index expression `x[hi:lo]`.{firrtl} is `UInt<hi - lo +
+1>`.{firrtl} (even if `x` is an `SInt`) and the type of `x[i]`.{firrtl} is
+`UInt<1>`.{firrtl}.
+
+The bit-index can be used as a sink or source. When used as a source,
```

Comment:
Have you considered restricting your proposal to L-value bit slices?
All R-value uses seem to already be covered by `bits` and thus it might help to focus the proposal on sub-word _assignments_.

Reply:
It's true that if it isn't restricted to L-value slices there is redundancy with the `bits` primop, but I think the intention is to move to a new syntax that is consistent for both L and R-values, and phase out the bits primop in the future. This also has the benefit of making the Chisel emission simpler, since it doesn't have to emit different syntax based on whether the index is an L-value or an R-value.

Reply:
> But I think the intention is to move to a new syntax that is consistent for both L and R-values, and phase out the bits primop in the future.

I don't think that is necessary. As I said, firrtl is an IR, so it can be more explicit about some things than a user-facing language. I am very much opposed to introducing duplicate functionality since it will make the compiler more complicated for little to no gain.

Reply:
> I am very much opposed to introducing duplicate functionality since it will make the compiler more complicated for little to no gain.

To clarify: the plan _is_ to remove `bits` for this exact reason and replace it with bit-index. 

Is the concern about atomicity of updates to the spec? I was planning to just remove `bits` in a separate PR in favor of bit index.

Reply:
> To clarify: the plan is to remove bits for this exact reason and replace it with bit-index.

Why not keep `bits`? As @darthscsi pointed out above, `bits` is easier to parse and does not conflict with sub-access. Otherwise to distinguish sub-access and bit-index, one would have to know the type of the expression.

> Is the concern about atomicity of updates to the spec? I was planning to just remove bits in a separate PR in favor of bit index.

Yes. I believe that before merging into `main`, the complete proposal should be considered.

In addition to that, removing `bits` will make for a back-wards incompatible change and I do not see a good reason for this, when we could just stick with the old syntax and either extend it to l-values or come up with an alternative syntax specifically for l-values. The handling of l-value and r-value bit-index is quite different anyways.

Reply:
> Why not keep bits? As @darthscsi pointed out above, bits is easier to parse and does not conflict with sub-access. Otherwise to distinguish sub-access and bit-index, one would have to know the type of the expression.

It seems clean to have a single unified op for extraction.

FIRRTL really screwed up here with not adding type information to each operation. Any sane parser is going to track the types of references and uses that to build up its internal FIRRTL IR (which necessarily must include type information). Hence, I'm not super concerned about this. I do admit that this means `foo[a] <= bar` is ambiguous without extra context. However, `foo[a] <= bar : uint<8>, uint<1>` is not and would be a great direction to that the FIRRTL textual format.

> Yes. I believe that before merging into main, the complete proposal should be considered.
>
> In addition to that, removing bits will make for a back-wards incompatible change and I do not see a good reason for this, when we could just stick with the old syntax and either extend it to l-values or come up with an alternative syntax specifically for l-values. The handling of l-value and r-value bit-index is quite different anyways.

I'm fine to just remove `bits` entirely here then.

I don't know if backwards compatibility should be a goal here. We're attempting to make it easy for FIRRTL compilers to check if they support a given FIRRTL text via: https://github.com/chipsalliance/firrtl-spec/pull/30. I guess my concern is that it seems weird to try to be backwards compatible for bits on the RHS of a connect when the fundamental change is to extend the spec in an entirely backwards incompatible way. Or: SFC will be "incompatible" with FIRRTL 2.0.0+ after this change even though it will work for Chisel designs where a user doesn't use bit index.

Reply:
> Or: SFC will be "incompatible" with FIRRTL 2.0.0+ after this change even though it will work for Chisel designs where a user doesn't use bit index.

My main point is that it will be harder to support older FIRRTL versions with the same compiler if we make this change. If we only add a new operation, then we trivially support older FIRRTL versions. However, if we replace `bits` with `[..]`, then we will still have to keep around the old code to support older FIRRTL versions. Not an insurmountable problem for sure. But there also does not seem to be a real upside to switching from `bits` to `[...]`.